contributors/devel/sig-architecture/api_changes.md

```diff
@@ -154,18 +159,42 @@ without changing the API version, so you can simply change it to:
 // Still API v6.
 type Frobber struct {
   Height int    `json:"height"`
-  Width  int    `json:"width"`
+  Width  *int   `json:"width"`  // optional, defaults to 42
   Param  string `json:"param"`
 }
 ```
 
+Note that fields which are added MUST be optional.  Optional fields must be
+nil-able (pointers or slices or (rarely) maps).
+
 The onus is on you to define a sane default value for `Width` such that rules
 #1 and #2 above are true - API calls and stored objects that used to work must
 continue to work.
 
+For newly added fields, the default behavior must be defined.  If this is a net
+new capability which did not exist before (i.e. a user must explicitly
+set it to change behavior), the default value should be "nothing" (`nil` in
+Go), meaning the feature is not activated. If this new field is extending an
+existing behavior, the default value should also be "nothing" (`nil`), meaning
+the existing behavior.  This satisfies rule #7.  If a user read the object
+before your feature and again after your feature, they would not see any
+change.
+
+NOTE: The default being `nil` is a change from previous guidance, and is
+counter to many previous changes.  The rationale for this is that clients (e.g.
+controllers) sometimes do things like hash objects and save that hash.  When a
```

Comment:
This is a controller bug that is possible no matter what we do to the API. Controllers need to hash exactly and only the fields they care about.

Actually I think a hash is almost certainly the wrong way to solve this problem. We give out UIDs in response to POST requests; controllers should record that and not do any hashing like this at all.

Reply:
> We give out UIDs in response to POST requests

a UID isn't unique to a particular write result... uid+resourceVersion (for the whole object) or uid+generation (for the spec) are more reasonable shorthands for "object X at state Y"

Reply:
I agree that this hashing is trouble, but I don't see how UID matters here?  E.g. in the case of deployment, status writes are fine but spec and metadata writes would need some reaction.

If we demand specific fields, we will certainly miss real changes because one of N open-coded field lists got missed..

Reply:
I mention UID because the problematic case has traditionally been controllers not recognizing the objects they create.

Reply:
And yes, comparing UID and generation is the canonical way to check if a spec has changed. RV change tells you something (spec or status) is changed.

Reply:
Generation is under manual control, right?  Not every resource has it?

Reply:
Can we find those places that are hashing and force them to use generation?

Reply:
Generation is in metadata and generally every resource sets it. It's not 100% automatic so it's possible some resources did it wrong, but I think we've found those cases by now. CRDs set generation automatically.

"We" can certainly find the controllers and fix them for some value of "we" :) Might be a good help-wanted thing for sig apps / api-machinery (individual controllers are usually not owned by api machinery).

Reply:
We can fix our own mess, anyway.

Reply:
A list of concrete examples would probably help direct the conversation. Laying out an example with podspecs:

* Controller manages a replicaset with podspec hash A
* Controller is upgraded to recognize previously unknown fields
* Controller re-derives hash for a replicaset's podspec, and it changes to B since it includes some unknown field

Assuming the field was defaulted and the no other writers, if the controller were to derive the hash on a pod that it had previously created, the hash will presumably match the B. The problem arises when controllers store A in a pod's labels and assume that its deterministically recomputable between versions. 

In that example, we are using hashes as cheap, derivable resource versions for subobjects but then we go and store the the version somewhere. It seems like we hit rough spots because we mix the use of stored and derived resource versions. We have not been able to keep stored and derived versions consistent over upgrades, and the proposal is one way to fix that but we're paving over some bad design. Two other approaches to consider:

* Fully derived approach: we stop storing derived resource versions. Versions are computed locally as needed. This makes selection based on version much more complex. To implement selection based on version, clients of the API would need to communicate the derivation (e.g. an ordered field mask to apply to an object before hashing) along with the resource version.
* Fully stored approach: controllers stop using hashes and move onto generations that gets rev'd every time some portion of an object is touched. This has some issues with version skew between controllers and the API and dropped fields. Controllers that could pass through unknown fields would help here.

Reply:
> Controller re-derives hash for a replicaset's podspec, and it changes to B since it includes some unknown field

Let me quote myself:

> Controllers need to hash exactly and only the fields they care about.

Reply:
That works with controllers that only care about a specific set of fields and never anymore but I don't see how that addresses the issue with the replicaset example. The controller needs to consider an increasing set of fields every upgrades. Options here are:
* This proposal: Make hash(podspec) accurately reflect that an object is unchanged. Defaults are always nil.
* Don't compare hash(podspec) that were derived from different versions of the API. This implies that we shouldn't store them in e.g. labels.
* Don't use hashes as a resource version. The fields a controller needs to hash may change, so they won't be stable.

Reply:
How can a controller which doesn't understand the contents of the template know whether a given change is important or not? The whole question is predicated on some changes being "unimportant" and not meriting a rollout. For this to make sense, a controller has to know something about the data and not treat it as an opaque blob.

There are ways to do this with hashes, but they aren't simple. Controllers who want to do this need to deliberately maintain the hashing function. Just JSON marshalling data that's not directly controlled by the controller code isn't good enough. The hash function really must not depend on e.g. the version of the client-go library you've linked against.

But if you do this, then there's an upgrade ordering problem. You have to upgrade the controller before the API, since the old controller won't understand the new field. But our current guidance is that the API schema gets upgraded first.

I think this entire approach is basically doomed.

A better approach might look like "take the current template, do a dry-run create, compare the result with a currently running pod, if there is a relevant difference, then do an update / rollout"

Reply:
Ya. I generally agree that a hash based approach is doomed. It's going to be broken or more complex and fragile than either generation or SSA approaches. Using generation has the downside that, right now, we bake them into the apiserver so they aren't particularly extensible, e.g. to custom resources, or to specific subsets of an object (although spec generation might be enough for most cases). I haven't thought through SSA yet, but it'd be interesting to see it proved out.